### DISCUSSION TITLE
Support for standalone Ring Panic button

### DISCUSSION THREAD

**Discussion by dochtie** [tsightler/ring-mqtt#883] [Feature Requests]
Is it possible to add support for the official ring panic button? I have looked through the discussions and can’t find a similar request.

I got the panic sliders from the keypad but have a standalone button for my child to press in a medical emergency. I have ring monitoring but it would be great to have HA integration.

**Comment by tsightler** [comment]
What exactly would you want to do with it?  There is nothing to automate with a button.

**Reply by dochtie** [reply]
I can think of a few different things. My use case is a medical emergency button. Right now the ring panic button calls the monitoring centre who then calls three emergency contact in order. 

I would automate the button to trigger the outside alarm (not the internal to freak my child out). As well as send snapshots to multiple people from the cameras.

Right now the button lights up when pressed but I’d like to have Alexa make an announcement so my son knows help is coming.

**Reply by tsightler** [reply]
But is what you are asking for actually possible?  It's not clear to me how the button works, I assumed it would simply trigger the respective alarm response, and that the state of the alarm control panel would change to some alarm state. Is that not the case?  I would have expected that you can trigger additional automations from the alarm control panel attributes in this case as I would exepcted at least the alarmState attribute would change from all-clear.

However, it almost sounds like you are saying you don't just want to perform additional action, but actually override the default Ring actions.  Does the button have some way to configure it to do nothing by default?  If not, I don't think you could override default Ring response, only add additional automations, but that should be possible already.

**Reply by dochtie** [reply]
I want to detect when the button is pressed and then run automations in addition to the ring call centre response.

I have enable_panic true in configuration but I only have fire and police entities and no medical panic entity.

When the alarm is in the medical state (either triggered by the physical panic button or the medical button on the keypad), there is no entity in HA that reflects this alarm state.

By default the medical state is a silent alarm and it contacts the monitoring centre. I would like to detect when the alarm is in this state, and have various automations trigger.

Ideally I’d like to detect when this specific button is pressed because then I know it’s activated by a child. But even if there is a way to detect when the alarm is in panic medical mode this would be a great addition.

**Reply by dochtie** [reply]
Apologies - realised I’m not being clear. When I push the button the ring goes into alarm state. The keypads say “medical alert active” but no sirens go off. The call centre gets an alert…

but critically the attribute AlarmState shows all-clear. It does not update… this is true for fire and police too. For me AlarmState always shows all-clear unless in full alarm and then the attribute changes to ‘burglar-alarm’.

Maybe I’m missing a way to detect this though?

**Reply by tsightler** [reply]
> I would like to detect when the alarm is in this state, and have various automations trigger.

Are you sure that the alarmState attribute of the info sensor does not change when the alarm is in medical alert state?  I'm quite surprised that this is the case.

To be completely honest, I think it is unlikely this will be implemented.  I would need to spend time to understand how the panic buttons work, what states they trigger, etc.  I don't own a panic button, don't plan to own one, and don't have time to dig into all of it.  Plus, testing all of that is challenging because, well, you have to keep triggering actual emergency events.

If course, the code is there, you are welcome to enhance it and submit PRs.

**Reply by dochtie** [reply]
Yea I’m absolutely sure. I put it into self monitoring mode and did some tests this morning. The attributes do not change (remains “all-clear”) when triggered from the keypad or the panic button.

If triggered from MQTT via the fire or police buttons I get burglar-accelerated-alarm or fire-accelerated alarm. 

A workaround for me would if a medical panic button trigger could be added via MQTT? Then I could program any smart button to trigger this mode, just like I can trigger fire or police.

**Reply by dochtie** [reply]
> > I would need to spend time to understand how the panic buttons work, what states they trigger, etc. I don't own a panic button, don't plan to own one, and don't have time to dig into all of it.

I completely understand. Just a request. I don’t have the technical skill unfortunately.

Thanks for an incredible tool.

**Reply by tsightler** [reply]
> Yea I’m absolutely sure. I put it into self monitoring mode and did some tests this morning. The attributes do not change (remains “all-clear”) when triggered from the keypad or the panic button.

I'm starting to wonder if this part is a bug.  Right now the code assumes that this value would not change unless the actual alarm mode also changed, but that's not really true in a medical alert case s I'm thinking it might be changing, but not being published. Would you be willing to test the dev branch with a small change to prove this?

Also, to support the button itself, would you be willing to run the Device Data Discovery process and send me the output via email (same username at gmail).  I would think supporting the button itself might be quite simple overall with this data. 
 https://github.com/dgreif/ring/wiki/Data-Discovery

**Reply by dochtie** [reply]
> I'm starting to wonder if this part is a bug. Right now the code assumes that this value would not change unless the actual alarm mode also changed, but that's not really true in a medical alert case s I'm thinking it might be changing, but not being published. Would you be willing to test the dev branch with a small change to prove this?

Yes absolutely! Thank you! The alarm mode doesn’t change - it remains in disarmed or home. So this could be it. I will download the dev branch now.

> Also, to support the button itself, would you be willing to run the Device Data Discovery process and send me the output via email (same username at gmail). I would think supporting the button itself might be quite simple overall with this data. https://github.com/dgreif/ring/wiki/Data-Discovery

Will do my best to run this now! Thank you. It’s the Keypad version 2 panic buttons and the standalone panic button.

**Reply by tsightler** [reply]
Well, I haven't made the change to the dev branch yet, so don't run it yet!  Also, you don't have to download anything, just set the branch config option to 'dev' and restart the addon (or BRANCH=dev environment variable for Docker).

But you don't want to run it yet, because the current dev branch implements the switch to new style FCM push notifications and there's no going back.  To be fair, this should be pushed to prod hopefully sometime this weekend, but once you move it's not possible to go back to prior versions.  Although, if your don't have cameras or intercoms, this is mostly irrelevant.

**Reply by dochtie** [reply]
Got it! thank you!

**Reply by dochtie** [reply]
> Also, to support the button itself, would you be willing to run the Device Data Discovery process and send me the output via email (same username at gmail). I would think supporting the button itself might be quite simple overall with this data. https://github.com/dgreif/ring/wiki/Data-Discovery

sent now :) Thank you again for taking a look into this.

**Reply by tsightler** [reply]
Make sure you don't have any offline devices, otherwise data discovery will hang forever (same for ring-mqtt startup as it waits for devices to be online).

BTW, out of curiosity, I'm assuming pressing the button generates some time of push notification in the Ring app?  I have plans to include push notification text as an attribute in a future version (maybe even 5.7.0 if I can sneak a few hours of time) so this might be a way to do this even without specific support for buttons.

**Reply by dochtie** [reply]
> BTW, out of curiosity, I'm assuming pressing the button generates some time of push notification in the Ring app? I have plans to include push notification text as an attribute in a future version (maybe even 5.7.0 if I can sneak a few hours of time) so this might be a way to do this even without specific support for buttons.

I ran it in Linux without issue. In windows it was hanging for some reason. Hopefully you got the email.

Pressing the button generates an alarm (albeit completely silent). All devices with the app get a critical alert just like the alarm was going off. 

Screenshot attached.
![IMG_9335](https://github.com/user-attachments/assets/e92ac2a5-20fc-48b7-a440-01c164e01612)

I also tested the buttons on the ring keypad 2 and looked closely at HA entities again. I can’t see any attribute change when the alarm is sounding in this way, not from the fire, police or medical buttons.

**Reply by tsightler** [reply]
I took a look at the output.  I see the buttons, but it's not clear to me how I could monitor for changes as there doesn't appear to be a state.  This isn't that unexpected as buttons are generally stateless and used for triggers, but the problem is all current devices are based on state updates.  If there is no state, I'm not sure how to monitor for it.

However, I'm really struggling to believe that when the alarm is in this state that there shouldn't be some kind of state change, it seems like there is a state change in the app, and that's the same API ring-mqtt is using.  Maybe I can simulate this with the built in panic sliders in the app, as I'm guess that triggers the same condition with the alarm.

**Reply by dochtie** [reply]
Thank you for looking into it.

I don’t have panic sliders in my app so that is interesting! 

Triggering any of the panic modes via the keypad or the panic buttons I don’t see any attribute change. Even when all my alarms are blaring for fire or police. 

If triggered using ring-mqtt then the attributes do update. But there isn’t a medical one I can use to trigger the medical response.

**Comment by tsightler** [comment]
@dochtie If you are still interested in this, would you be willing to test a development version?  I haven't done it yet, but I was thinking I might add some dummy code that at least logs and data updates when a button is pressed so maybe we can figure out what data to trigger on.  However, since I don't have a button, I'd need someone willing to test and share the logs back to me.

**Reply by dochtie** [reply]
> @dochtie If you are still interested in this, would you be willing to test a development version? I haven't done it yet, but I was thinking I might add some dummy code that at least logs and data updates when a button is pressed so maybe we can figure out what data to trigger on. However, since I don't have a button, I'd need someone willing to test and share the logs back to me.

Absolutely- let me know what you need. Thank you!

**Reply by tsightler** [reply]
I have pushed some code to the dev branch that should at least discover the panic buttons during startup and should log any messages/data updates they generate.  Ideally if you could test the dev branch (just set "branch" config option to "dev" and restart, it will pull down the latest dev branch on startup).  You should see the panic button during discovery and, if you could press the button and note if it logs anything, I need the JSON data that is logged in that case.
